process/block/shardblock.go

```diff
@@ -473,32 +505,51 @@ func (sp *shardProcessor) CommitBlock(
 		return err
 	}
 
-	_, err = sp.accounts.Commit()
-	if err != nil {
-		return err
-	}
-
-	sp.blocksTracker.AddBlock(header)
+	maxNonce := sp.getMaxNonceFromUsedMetaBlocks(header)
 
-	log.Info(fmt.Sprintf("shardBlock with nonce %d and hash %s has been committed successfully\n",
+	log.Info(fmt.Sprintf("the highest metachain block nonce used in shard header with nonce %d is %d\n",
 		header.Nonce,
-		core.ToB64(headerHash)))
+		maxNonce))
 
-	errNotCritical = sp.txPreProcess.RemoveTxBlockFromPools(body, sp.dataPool.MiniBlocks())
+	processedMetaHdrs, errNotCritical := sp.getProcessedMetaBlocksFromPool(body, maxNonce)
 	if errNotCritical != nil {
 		log.Debug(errNotCritical.Error())
 	}
 
-	processedMetaHdrs, errNotCritical := sp.getProcessedMetaBlocksFromPool(body)
+	errNotCritical = sp.saveLastNotarizedHeader(sharding.MetachainShardId, processedMetaHdrs)
 	if errNotCritical != nil {
 		log.Debug(errNotCritical.Error())
 	}
 
-	err = sp.saveLastNotarizedHeader(sharding.MetachainShardId, processedMetaHdrs)
+	sp.mutNotarizedHdrs.RLock()
+	lastNotarizedNonce := sp.lastNotarizedHdrs[sharding.MetachainShardId].GetNonce()
+	sp.mutNotarizedHdrs.RUnlock()
+
+	log.Info(fmt.Sprintf("last notarized block nonce from metachain is %d and max nonce used in current header is %d\n",
+		lastNotarizedNonce,
+		maxNonce))
+
+	if maxNonce > 0 && lastNotarizedNonce != maxNonce {
```

Comment:
add this verification inside the saveLastNotarizedHeader function and make it return a critical error.

Reply:
Done